### ISSUE TITLE
fts_fuzzy_match: Recursion count doesn't get decremented anywhere

### ISSUE THREAD

**Issue by arch4672** [forrestthewoods/lib_fts#13]
The recursionCount variable can reach the recursionLimit very quickly as it is not being calculated correctly.   It means in some cases not all of the possible matches get exhaustively searched.

I think recursionCount is meant to count at what level of recursion you're currently at, but because it doesn't get decremented once a recursive call has exited it's not correct. 

I think decrementing recursionCount after the call to fuzzy_match_recursive() fixes the problem.

Thanks for sharing this library by the way - it's really useful.

**Comment by forrestthewoods** [comment]
> It means in some cases not all of the possible matches get exhaustively searched.

That is my intent. My goal is not to limit recursion depth to prevent stack overflows. My goal is to keep fuzzy_match reliably performant. I don't want degenerate cases such as fuzzy_match("aaaaaaaaa", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa") to freeze the application.

It's sacrificing a small amount of accuracy for more consistent performance. I think that's a fair trade-off.

Do you have a particular use case where this is causing you a problem?

**Comment by arch4672** [comment]
Fair enough. For my application, the chances of having an item with the text "aaaaaaaaaaaaaaaaaaa" are pretty slim, so I've made recursionCount in my version count the recursion depth.

The use case that pointed me to this was if you searched for "CH CBN DECK SPKR SUPP" with "decksupp".

It found the item, but didn't give it the highest score possible as it matched it like this (where matches are between { }'s):

"CH CDB {DECK} {S}PKR S{UPP}"

The highest scoring match is actually:

"CH CDB {DECK} SPKR {SUPP}"

Modifying your code so it decrements the recusion counter sorted this out for me.  I thought I'd point it out, but it sounds like your code is doing what you intended.

**Comment by forrestthewoods** [comment]
Interesting test case. I'll try to take a closer look using it. Thanks.

**Comment by octavian-nita** [comment]
Indeed :)

You can also notice it in your [demo](https://s3-us-west-2.amazonaws.com/forrestthewoods.staticweb/lib_fts/tests/fuzzy_match/fts_fuzzy_match_test.html), when searching the Unreal Engine 4 Filenames using the ```Mock``` pattern and ordering by Best Match... The first several matches now look like:

21 - MockAI.h
20 - MacroCallback.h
19 - MockAI.cpp
18 - MacroCallback.cpp
18 - MockAI_BT.h
16 - MockAI_BT.cpp
13 - MouseClickProcessor.cs
13 - MovieSceneColorTrack.h
11 - MockGameplayTasks.h
11 - MovieSceneColorTrack.cpp

and I would say ```MockAI.cpp``` is a better match than ```MacroCallback.h``` :)

It looks like proximity-from-the-start should be overridden by something like proximity-of-range-matched???

It's true that the demo still uses the older, non-exhaustive JavaScript search but when running the C++ version on similar inputs I got something very similar.

**Comment by bubnikv** [comment]
@arch4672 is right about his case. The recursion count accumulates pretty quickly over 10 the way the current code works and in the case given by @arch4672 the maximum "recursion" is reached before the 2nd match.

I understand the intent, but the implementation is not quite right.

**Comment by forrestthewoods** [comment]
> I understand the intent, but the implementation is not quite right.

@bubnikv, Agreed.

When I did my original blog post in 2016 there was nothing public in the way of "sublime fuzzy matching". I took a bit of a stab in the dark. 

Does anyone know what the latest and greatest in fuzzy matching algorithms is? There's been a definite increase in tools leverages fuzzy search over the past few years. But I'm not sure what algorithms are out there.

**Comment by bubnikv** [comment]
I found one library with a summary of other fuzzy search algorithms.
https://github.com/jhawthorn/fzy/blob/master/ALGORITHM.md

I would love to review all those algorithms, but unfortunately I don't have
so much time on my hands. I plan to tweak the weights of your algorithm and
to add some non-linear increase of match with number of successive
character matches.
